src/powershell/Program.cs

```diff
@@ -19,7 +63,458 @@ public sealed class ManagedPSEntry
+                    // | exec_path | ... \0
+                    // | argv[0]   | ... \0
+                    // | argv[1]   | ... \0
+                    //   ...
+                    //
+                    // We care about argv[0], since that's the name the process was started with.
+                    // If argv[0][0] == '-', we have been invoked as login.
+                    // Doing this, the buffer we populated also recorded `exec_path`,
+                    // which is the path to our executable `pwsh`.
+                    // We can reuse this value later to prevent needing to call a .NET API
+                    // to generate our exec invocation.
+
+
+                    // We don't care about argc's value, since argv[0] must always exist.
+                    // Skip over argc, but remember where exec_path is for later
+                    execPathPtr = IntPtr.Add(procargs, sizeof(int));
+
+                    // Skip over exec_path
+                    byte *argvPtr = (byte *)execPathPtr;
+                    while (*argvPtr != 0) { argvPtr++; }
+                    while (*argvPtr == 0) { argvPtr++; }
+
+                    // First char in argv[0]
+                    procNameFirstByte = *argvPtr;
+                }
+
+                if (!IsLogin(procNameFirstByte, args))
+                {
+                    return;
+                }
+
+                // Get the pwshPath from exec_path
+                pwshPath = Marshal.PtrToStringAnsi(execPathPtr);
+
+                // exec pwsh
+                ThrowOnFailure("exec", ExecPwshLogin(args, pwshPath, isMacOS: true));
+            }
+            finally
+            {
+                Marshal.FreeHGlobal(procargs);
+            }
+        }
+
+        /// <summary>
+        /// Checks args to see if -Login has been specified.
+        /// </summary>
+        /// <param name="procNameFirstByte">The first byte of the name of the currently running process.</param>
+        /// <param name="args">Arguments passed to the program.</param>
+        /// <returns></returns>
+        private static bool IsLogin(
+            byte procNameFirstByte,
+            string[] args)
+        {
+            // Process name starting with '-' means this is a login shell
+            if (procNameFirstByte == 0x2D)
+            {
+                return true;
+            }
+
+            // Check the parameters in order
+            for (int i = 0; i < args.Length; i++)
+            {
+                string arg = args[i];
+
+                if (arg == null || arg.Length < 2)
+                { 
+                    continue;
+                }
+
+                // '/?' is equivalent to '-help'
+                if (arg.Equals("/?"))
+                {
+                    return false;
+                }
+
+                // Must look like '-<name>'
+                if (arg[0] != '-')
+                {
+                    continue;
+                }
```

Comment:
Think about this: `pwsh script.ps1 -log`. With the current code, we will treat it incorrectly as a login shell.